packages/server/graphql/mutations/setPokerSpectate.ts

```diff
@@ -47,6 +50,9 @@ const setPokerSpectate = {
         .get(teamMemberId)
         .update({isSpectatingPoker: isSpectating, updatedAt: now})
     }).run()
+
+    // mutate the dataLoader cache
+    meetingMember.isSpectating = isSpectating
```

Comment:
Use a fresh DataLoader or call dataLoader.clear() can fix this bug. But when I check other mutations such as `renameMeeting`, it doesn't need to neither update nor clear the cache, it also works well.
Do you know the root cause? @Dschoordsch

Reply:
I got other two questions. Begging for your help.
1. I try to remove `did` from `subOptions` in the mutation, but it still get the same `dataLoader` in subscriptions.
2. When I remove `publish(....)` from mutation, everything works fine. Is it redundant?
Thx!

Reply:
@seabornlee the `operationId` in `subOptions` is passed on as `dataloaderId` here https://github.com/ParabolInc/parabol/blob/f9c00c48cb8600bff2f8e3040a77580f006be45a/packages/server/graphql/ResponseStream.ts#L23 so you would need to remove that to test it. But we pass it intentionally to reduce the DB roundtrips for all subscriptions. Publish is not redundant since it notifies all meeting participants, who would otherwise have stale data.
We always have to update or clear the dataloader if we modify an entry in DB after we've loaded it through the dataloader and we pass the dataloader to the subscription. Skimming through the mutations it seems most of the time we're not publishing the same entities we've loaded before, that's why it's not an issue most of the time.

Reply:
I see, so the sequence of events with the previous implementation was:
- call `SetSpectateMutation` which creates a new dataLoader and cache
- in the mutation, grab the `meetingMember` with the dataLoader
- update the `meetingMember` in the db
- subscription then uses the stale dataLoader when reading the `meetingMember`
- call `SetSpectateMutation` again, which creates another dataLoader and cache using the previously updated `meetingMember`
- now the subscription will read the `meetingMember` that was updated in the previous mutation execution

That makes sense, but it feels error-prone. I wonder whether we could clear the dataLoader automatically after executing a mutation?

Reply:
That's the trade of we're making for performance. It's specifically designed so that subscriptions reuse the data and don't query the db again.
It seems like it doesn't occur that often, but we could add a verification mode, where instead of reusing the dataloader it would create a new one and compare it. This we could activate for dev and staging to find these errors earlier.

Reply:
Makes sense, and thanks for creating the issue. I look forward to getting Matt's thoughts on this. The same stale dataLoader bug did crop up in a recent PR here: https://github.com/ParabolInc/parabol/pull/5674